### PR TITLE
objstr: Refactored the 'mul' method

### PR DESCRIPTION
**Pull request by ZapAnton**
- Used `str::repeat` instead of manually building the result string
- Rewritten the int type check to return error at the beginning of the
method
- Replaced the `unwrap` calls with the single `unwrap_or`

### CODE REVIEW

File: vm/src/obj/objstr.rs [hunk 1]
```diff
@@ -183,22 +183,16 @@ impl PyString {
...
-            Ok(result)
-        } else {
-            Err(vm.new_type_error(format!("Cannot multiply {} and {}", self, val)))
+        if !objtype::isinstance(&val, &vm.ctx.int_type()) {
+            return Err(vm.new_type_error(format!("Cannot multiply {} and {}", self, val)));
         }
+        let value = &self.value;
+        let multiplier = objint::get_value(&val)
+            .to_i32()
+            .map(|multiplier| if multiplier < 0 { 0 } else { multiplier })
```
Comment (sapir) [inline vm/src/obj/objstr.rs]: you could use `multiplier.max(0)`

Comment (ZapAnton) [reply vm/src/obj/objstr.rs]: Done

File: vm/src/obj/objstr.rs [hunk 2]
```diff
@@ -183,22 +183,16 @@ impl PyString {
...
-        } else {
-            Err(vm.new_type_error(format!("Cannot multiply {} and {}", self, val)))
+        if !objtype::isinstance(&val, &vm.ctx.int_type()) {
+            return Err(vm.new_type_error(format!("Cannot multiply {} and {}", self, val)));
         }
+        let value = &self.value;
+        let multiplier = objint::get_value(&val)
+            .to_i32()
+            .map(|multiplier| multiplier.max(0))
+            .and_then(|multiplier| multiplier.to_usize())
```
Comment (coolreader18) [inline vm/src/obj/objstr.rs]:
```suggestion
            .and_then(i32::to_usize)
```

**Comment by silmeth** [comment]
You could also switch from `i32` to `isize` and return an `OverflowError` instead of doing `unwrap_or(0)` if the multiplier doesn’t fit in, it seems that’s what CPython does:

```python
>>> 'abc' * 234234234234234234234234234234 # does not fit in i64
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
OverflowError: cannot fit 'int' into an index-sized integer
>>> 'a' * 2147483650 # does not fit in i32, but works on 64 bit machines
'aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa… # etc.
```

File: vm/src/obj/objstr.rs (hunk 2 above)
Comment (ZapAnton) [reply vm/src/obj/objstr.rs]: Got an error:
```text
expected signature of `fn(i32) -> _`
found signature of `for<'r> fn(&'r _) -> _`
```

Comment (coolreader18) [reply vm/src/obj/objstr.rs]: Oh, it's fine then.

**Pull request merged**